app/models/order_detail/notices.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module OrderDetail::Notices
+  extend ActiveSupport::Concern
+
+  included do
+    serialize :notice_keys, Array
+    serialize :problem_keys, Array
+
+    before_save :set_problem_and_notices
+    before_save :update_fulfilled_at_on_resolve
+  end
+
+  def set_problem_and_notices
+    notice_service = OrderDetails::NoticesService.new(self)
+
+    self.problem_keys = notice_service.problems
+    self.notice_keys = notice_service.notices
+    self.problem = problem_keys.present?
+  end
+
+  # This must be called after updating problem field
+  def update_fulfilled_at_on_resolve
```

Comment:
`set_problem_order` had `true # problem might be false; we need the callback chain to continue` as the last line. Should we also do that here?

Reply:
I've found no documentation on the return value affecting the callback chain. It's only interrupted by an exception or throwing :abort. https://guides.rubyonrails.org/v7.0/active_record_callbacks.html#halting-execution